VotingApplication/VotingApplication.Web/Api/Controllers/PollResultsController.cs

```diff
@@ -105,31 +84,46 @@ private static DateTime UnixTimeToDateTime(double unixTimestamp)
             return dateTime;
         }
 
-        private static ResultsRequestResponseModel GenerateResults(IEnumerable<Vote> votes, bool namedVoting)
+        private static ResultsRequestResponseModel GenerateResults(IEnumerable<Vote> votes, IEnumerable<Choice> choices, bool namedVoting)
         {
-            var groupedResults = (from vote in votes
-                                  group vote by vote.Choice.PollChoiceNumber into result
-                                  let optionGroupVotes = result.ToList()
-                                  select new
-                                             {
-                                                 ChoiceName = optionGroupVotes[0].Choice.Name,
-                                                 Sum = optionGroupVotes.Sum(v => v.VoteValue),
-                                                 Voters = optionGroupVotes.Select(v => CreateResultVoteModel(v, namedVoting)).ToList()
-                                             })
-                  .ToList();
-
-            if (!groupedResults.Any())
+            if (choices.Count() == 0)
+            {
+                return new ResultsRequestResponseModel
+                {
+                    Winners = new List<string>(),
+                    Results = new List<ResultModel>()
+                };
+            }
+
+            List<ResultModel> groupedResults = new List<ResultModel>();
+
+            foreach (Choice choice in choices)
```

Comment:
There's multiple enumeration going on here (`foreach` here and the `.Count()` above). The collection shouldn't be changing on enumeration, but there's nothing to prevent it happening with a different collection object.